library/core/src/intrinsics/simd.rs

```diff
@@ -0,0 +1,412 @@
+//! SIMD compiler intrinsics.
+//!
+//! In this module, a "vector" is any `repr(simd)` type.
+
+extern "platform-intrinsic" {
+    /// Add two simd vectors elementwise.
+    ///
+    /// `T` must be a vector of integer or floating point primitive types.
+    pub fn simd_add<T>(x: T, y: T) -> T;
+
+    /// Subtract `rhs` from `lhs` elementwise.
+    ///
+    /// `T` must be a vector of integer or floating point primitive types.
+    pub fn simd_sub<T>(lhs: T, rhs: T) -> T;
+
+    /// Multiply two simd vectors elementwise.
+    ///
+    /// `T` must be a vector of integer or floating point primitive types.
+    pub fn simd_mul<T>(x: T, y: T) -> T;
+
+    /// Divide `lhs` by `rhs` elementwise.
+    ///
+    /// `T` must be a vector of integer or floating point primitive types.
+    ///
+    /// # Safety
+    /// For integers, `rhs` must not contain any zero elements.
+    /// Additionally for signed integers, `<int>::MIN / -1` is undefined behavior.
+    pub fn simd_div<T>(lhs: T, rhs: T) -> T;
+
+    /// Remainder of two vectors elementwise
+    ///
+    /// `T` must be a vector of integer or floating point primitive types.
+    ///
+    /// # Safety
+    /// For integers, `rhs` must not contain any zero elements.
+    /// Additionally for signed integers, `<int>::MIN / -1` is undefined behavior.
+    pub fn simd_rem<T>(lhs: T, rhs: T) -> T;
+
+    /// Elementwise vector left shift.
+    ///
+    /// Shift `lhs` left by `rhs`, shifting in sign bits for signed types.
+    ///
+    /// `T` must be a vector of integer primitive types.
+    ///
+    /// # Safety
+    ///
+    /// Each element of `rhs` must be less than `<int>::BITS`.
+    pub fn simd_shl<T>(lhs: T, rhs: T) -> T;
+
+    /// Elementwise vector right shift.
+    ///
+    /// Shift `lhs` right by `rhs`, shifting in sign bits for signed types.
+    ///
+    /// `T` must be a vector of integer primitive types.
+    ///
+    /// # Safety
+    ///
+    /// Each element of `rhs` must be less than `<int>::BITS`.
+    pub fn simd_shr<T>(lhs: T, rhs: T) -> T;
+
+    /// Elementwise vector "and".
+    ///
+    /// `T` must be a vector of integer primitive types.
+    pub fn simd_and<T>(x: T, y: T) -> T;
+
+    /// Elementwise vector "or".
+    ///
+    /// `T` must be a vector of integer primitive types.
+    pub fn simd_or<T>(x: T, y: T) -> T;
+
+    /// Elementwise vector "exclusive or".
+    ///
+    /// `T` must be a vector of integer primitive types.
+    pub fn simd_xor<T>(x: T, y: T) -> T;
+
+    /// Numerically cast a vector, elementwise.
+    ///
+    /// When casting floats to integers, the result is truncated.
+    /// When casting integers to floats, the result is rounded.
+    /// Otherwise, truncates or extends the value, maintaining the sign for signed integers.
+    ///
+    /// `T` and `U` be a vectors of integer or floating point primitive types, and must have the
+    /// same length.
```

Comment:
```suggestion
    /// `T` and `U` must be vectors of integer or floating point primitive types, and must have the
    /// same length.
```
Also IMO it'd make more sense to state the precondition first, since the behavior description is basically doing case distinction on this preconditions